### DISCUSSION TITLE
Understanding subtyping matching

### DISCUSSION THREAD

**Discussion by CharlieTap** [WebAssembly/gc#539] [Q&A]
Hello,

I'm running into a problem when testing my runtime against the wasm test suite, specifically this test:
https://github.com/WebAssembly/gc/blob/e5af64983adf256f3472de4e4e6b81e5a0063c82/test/core/gc/ref_test.wast#L276

I'm currently matching these two types as I've made the following assumptions:

- Both references are non null
- Reference 1 is a reference to the abstract heap type struct
- Reference 2 is a defined type which when expanded is a composite type struct

Looking at the matching logic from the reference interpreter it would appear that these should match:

https://github.com/WebAssembly/gc/blob/e5af64983adf256f3472de4e4e6b81e5a0063c82/interpreter/valid/match.ml#L78

So I'm guessing maybe Reference 1 should not be a abstract heap type of struct? I get this after substituting the reference type embedded in the instruction 🤔 

Somewhere I'm going wrong and I feel like I should be comparing two defined types and thus would be not matching because the structs don't match

**Comment by tlively** [comment]
Reference 1 is also a defined type, not the abstract heap type `struct`. `(type $t0 (sub (struct)))` is a defined struct type with zero fields.

**Reply by CharlieTap** [reply]
Thank you, I think its my type_of_ref equivalent function should be more complex for the Function, Array and Struct scenarios. 

For anyone looking at this in the future, the logic for these branches appears:

Struct and Arrays

https://github.com/WebAssembly/gc/blob/e5af64983adf256f3472de4e4e6b81e5a0063c82/interpreter/runtime/aggr.ml#L59

Functions

https://github.com/WebAssembly/gc/blob/e5af64983adf256f3472de4e4e6b81e5a0063c82/interpreter/runtime/instance.ml#L43
